util/Pos/Util/Chrono.hs

```diff
@@ -53,6 +57,9 @@ makeWrapped ''OldestFirst
 deriving instance Semigroup (f a) => Semigroup (NewestFirst f a)
 deriving instance Semigroup (f a) => Semigroup (OldestFirst f a)
 
+deriving instance Monoid (f a) => Monoid (NewestFirst f a)
+deriving instance Monoid (f a) => Monoid (OldestFirst f a)
```

Comment:
These instances are going to break the invariant that `NewestFirst` etc want to preserve by default. You need a precondition on this that, for any `as <> bs`, every element in `bs` is older than any element in `as`.

The semigroup instance would need to do sorting on the result list, or some kind of merge sort, but then you need a way to project the chronological aspect that you're sorting on.